### DISCUSSION TITLE
Why do you choose Rust to build the DSL

### DISCUSSION THREAD

**Discussion by Z-KN** [cucapra/futil#1306] [Q&A] [answered]
Dear authors,

I am impressed by your great work. There is one question in my mind: why do you use Rust to build the DSL? As far as I know, Calyx is the first IL built on Rust. Some other DSLs are built on C++ (Halide) or Scala (Chisel) or Python (MyHDL). Out of what kind of consideration did you use Rust to write this program (IL)?

I want to do some similar work and hope you can share some insights with me.

**Comment by rachitnigam** [comment, marked as answer]
We wanted something that allowed us to build a fast and efficient tool chain and rust seemed like a saner alternative to c++. Overall, we’ve realtors liked Rust’s flexibility and efficiency!

**Reply by sampsyo** [reply]
Very true! To add one more dimension to your question: the other examples you cited (Halide, Chisel, and MyHDL) are all *embedded* DSLs. So a different question you may have asked might have been, "Why did you decide to design a new language (with a parser) instead of an embedded DSL ('just' a library in a sufficiently flexible general-purpose language)?"

There are many possible reasons for that, of course, but a main one is that Calyx is an IR: it's meant to be transformed, i.e., emitted programmatically. It would be really annoying if we had to consume and emit valid C++ or Scala code in every single compiler pass!

**Comment by Z-KN** [comment]
Thank you very much! I am surprised and grateful that you give me different aspects of methodology of building this work. I really benefit from it a lot. Wish Caylx a great success then!

Cheers!
